File: src/go/p4/search_insert_delete.go
package main

import (
	"fmt"
	"time"
	"sync"
	"math/rand"
	"container/list"
	"go/shared/parse"
)

type container struct {
	
	//Synchronization Members.
	deleteLock sync.RWMutex	//Used to exclude deletions from each other, and every other operation, while allowing the other operations to proceed in parallel.
	insertLock sync.Mutex	//Used to exclude more than one inserter from inserting at the same time.
	sizeLock sync.Mutex		//Used to determine the size of the list without torn reads.
	
	//Mutable Member.
	ctnr *list.List
	
}

func (c *container) find(x int) *(list.Element) {
	c.sizeLock.Lock()
	size := c.ctnr.Len()
	c.sizeLock.Unlock()
	
	for i, item := 0, c.ctnr.Front(); i < size; i++ {
		if item.Value == x {
			return item
		}
		item = item.Next()
	}
	return nil
}

func searcher(id int, c *container, done chan<- bool) {
	defer func() {done <- true}()
	
	time.Sleep(time.Duration(rand.Intn(1)) * time.Millisecond)
	//start := time.Now()
	
	c.deleteLock.RLock()
	defer c.deleteLock.RUnlock()
	
	if c.find(id) != nil {
		fmt.Printf("(Searcher %d) Found element {%d}.\n", id, id)
	}else{
		fmt.Printf("(Searcher %d) Did not find element {%d}!\n", id, id)
	}
	
	//fmt.Printf("%d\n", time.Since(start).Nanoseconds())
}

func inserter(id int, c *container, done chan<- bool) {
	defer func() {done <- true}()
	
	time.Sleep(time.Duration(rand.Intn(1)) * time.Millisecond)
	//start := time.Now()
	
	c.deleteLock.RLock()
	defer c.deleteLock.RUnlock()
	
	c.insertLock.Lock()
	defer c.insertLock.Unlock()
	
	func (){
		c.sizeLock.Lock()
		defer c.sizeLock.Unlock()
		
		c.ctnr.PushBack(id)
	}()
	
	//fmt.Printf("%d\n", time.Since(start).Nanoseconds())
	
	fmt.Printf("(Inserter %d) Added element {%d}.\n", id, id)
}

func deleter(id int, c *container, done chan<- bool) {
	defer func() {done <- true}()
	
	time.Sleep(time.Duration(rand.Intn(1)) * time.Millisecond)
	//start := time.Now()
	
	c.deleteLock.Lock()
	defer c.deleteLock.Unlock()
	
	if elem := c.find(id); elem != nil {
		c.ctnr.Remove(elem)
		fmt.Printf("(Deleter %d) Removed element {%d}.\n", id, id)
	}else{
		fmt.Printf("(Deleter %d) Did not find element {%d}!\n", id, id)
	}
	
	//fmt.Printf("%d\n", time.Since(start).Nanoseconds())
}

func testScenario(totalSearchers, totalInserters, totalDeleters int) {
	theContainer := container{}
	theContainer.ctnr = list.New()
	done := make(chan bool)
	
	for i, j, k := 0, 0, 0; i + j + k < totalSearchers + totalInserters + totalDeleters; {
		//time.Sleep(time.Duration(rand.Intn(5)) * time.Millisecond)
		time.Sleep(time.Duration(rand.Intn(1)) * time.Millisecond)
		if i < totalSearchers && j < totalInserters && k < totalDeleters {
			if rand.Intn(3) == 0 {
				go searcher(i, &theContainer, done)
				i++
			}else{
				if rand.Intn(2) == 0 {
					go inserter(j, &theContainer, done)
					j++
				}else{
					go deleter(k, &theContainer, done)
					k++
				}
			}
		}else if i < totalSearchers && j < totalInserters {
			if rand.Intn(2) == 0 {
				go searcher(i, &theContainer, done)
				i++
			}else{
				go inserter(j, &theContainer, done)
				j++
			}
		}else if j < totalInserters && k < totalDeleters {
			if rand.Intn(2) == 0 {
				go inserter(j, &theContainer, done)
				j++
			}else{
				go deleter(k, &theContainer, done)
				k++
			}
		}else if i < totalSearchers && k < totalDeleters {
			if rand.Intn(2) == 0 {
				go searcher(i, &theContainer, done)
				i++
			}else{
				go deleter(k, &theContainer, done)
				k++
			}
		}else if i < totalSearchers {
			go searcher(i, &theContainer, done)
			i++
		}else if j < totalInserters {
			go inserter(j, &theContainer, done)
			j++
		}else if k < totalDeleters {
			go deleter(k, &theContainer, done)
			k++
		}
	}
	
	for i := 0; i < totalSearchers + totalInserters + totalDeleters; i++ {
		<- done
	}
}

func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	
	fmt.Print("Please input how many searcher threads to run: ")
	if searchers, searchersErr := parse.ScanInt(); !searchersErr && searchers >= 0 {
		fmt.Print("Please input how many inserter threads to run: ")
		if inserters, insertersErr := parse.ScanInt(); !insertersErr && inserters >= 0 {
			fmt.Print("Please input how many deleter threads to run: ")
			if deleters, deletersErr := parse.ScanInt(); !deletersErr && deleters >= 0 {
				testScenario(searchers, inserters, deleters)
			}else{
				fmt.Println("Please input a single natural number, and nothing else.")
			}
		}else{
			fmt.Println("Please input a single natural number, and nothing else.")
		}
	}else{
		fmt.Println("Please input a single natural number, and nothing else.")
	}
}

File: src/cpp/p4/search_insert_delete.cpp
#include <list>
#include <thread>
#include <chrono>
#include <vector>
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <functional>
#include <shared_mutex>
#include "cpp/shared/parse.hpp"

typedef std::chrono::steady_clock testing_clock;

std::mutex output_mutex;

struct container{
	
	//Constructors/Destructor.
	container() : delete_lock(), insert_lock(), size_lock(), ctnr() {}
	container(const container&) = delete;
	container(container&&) = delete;
	~container() = default;
	
	//Assignment Operators.
	container& operator=(const container&) = delete;
	container& operator=(container&&) = delete;
	
	//Container Functions.
	std::list<int>::iterator find(int x);
	
	//Synchronization Members.
	std::shared_mutex delete_lock;
	std::mutex insert_lock;
	std::mutex size_lock;
	
	//Mutable Members.
	std::list<int> ctnr;

};

std::list<int>::iterator container::find(int x){
	size_lock.lock();
	int size = ctnr.size();
	size_lock.unlock();
	
	auto item = ctnr.begin();
	for(int i = 0; i < size; ++i){
		if(*item == x){
			return item;
		}
		++item;
	}
	throw std::range_error("The element is not in the list.");
}

void searcher(int id, container& c){
	std::this_thread::sleep_for(std::chrono::milliseconds(1));
	//testing_clock::time_point start = testing_clock::now();
	
	std::shared_lock<std::shared_mutex> del_lk(c.delete_lock);
	
	try{
		c.find(id);
		
		output_mutex.lock();
		std::cout << "(Searcher " << id << ") Found element {" << id << "}.\n";
		output_mutex.unlock();
	}catch(const std::range_error& ex){
		output_mutex.lock();
		std::cout << "(Searcher " << id << ") Did not find element {" << id << "}!\n";
		output_mutex.unlock();
	}
	
	/*output_mutex.lock();
	std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(testing_clock::now() - start).count() << "\n";
	output_mutex.unlock();*/
}

void inserter(int id, container& c){
	std::this_thread::sleep_for(std::chrono::milliseconds(1));
	//testing_clock::time_point start = testing_clock::now();
	
	std::shared_lock<std::shared_mutex> del_lk(c.delete_lock);
	std::unique_lock ins_lk(c.insert_lock);
	
	c.size_lock.lock();
	c.ctnr.push_back(id);
	c.size_lock.unlock();
	
	/*output_mutex.lock();
	std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(testing_clock::now() - start).count() << "\n";
	output_mutex.unlock();*/
	
	output_mutex.lock();
	std::cout << "(Inserter " << id << ") Added element {" << id << "}.\n";
	output_mutex.unlock();
}

void deleter(int id, container& c){
	std::this_thread::sleep_for(std::chrono::milliseconds(1));
	//testing_clock::time_point start = testing_clock::now();
	
	std::unique_lock<std::shared_mutex> del_lk(c.delete_lock);
	
	try{
		std::list<int>::iterator elem = c.find(id);
		c.ctnr.erase(elem);
		
		output_mutex.lock();
		std::cout << "(Deleter " << id << ") Removed element {" << id << "}.\n";
		output_mutex.unlock();
	}catch(const std::range_error& ex){
		output_mutex.lock();
		std::cout << "(Deleter " << id << ") Did not find element {" << id << "}!\n";
		output_mutex.unlock();
	}
	
	/*output_mutex.lock();
	std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(testing_clock::now() - start).count() << "\n";
	output_mutex.unlock();*/
}

void test_scenario(int total_searchers, int total_inserters, int total_deleters){
	container the_container;
	
	std::vector<std::thread> searchers;
	std::vector<std::thread> inserters;
	std::vector<std::thread> deleters;
	
	for(int i = 0, j = 0, k = 0; i + j + k < total_searchers + total_inserters + total_deleters; ){
		//std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 5));
		std::this_thread::sleep_for(std::chrono::milliseconds(1));
		if(i < total_searchers && j < total_inserters && k < total_deleters){
			if(std::rand() % 3 == 0){
				searchers.push_back(std::thread(searcher, i++, std::ref(the_container)));
			}else{
				if(std::rand() % 2 == 0){
					inserters.push_back(std::thread(inserter, j++, std::ref(the_container)));
				}else{
					deleters.push_back(std::thread(deleter, k++, std::ref(the_container)));
				}
			}
		}else if(i < total_searchers && j < total_inserters){
			if(std::rand() % 2 == 0){
				searchers.push_back(std::thread(searcher, i++, std::ref(the_container)));
			}else{
				inserters.push_back(std::thread(inserter, j++, std::ref(the_container)));
			}
		}else if(i < total_searchers && k < total_deleters){
			if(std::rand() % 2 == 0){
				searchers.push_back(std::thread(searcher, i++, std::ref(the_container)));
			}else{
				deleters.push_back(std::thread(deleter, k++, std::ref(the_container)));
			}
		}else if(j < total_inserters && k < total_deleters){
			if(std::rand() % 2 == 0){
				inserters.push_back(std::thread(inserter, j++, std::ref(the_container)));
			}else{
				deleters.push_back(std::thread(deleter, k++, std::ref(the_container)));
			}
		}else if(i < total_searchers){
			searchers.push_back(std::thread(searcher, i++, std::ref(the_container)));
		}else if(j < total_inserters){
			inserters.push_back(std::thread(inserter, j++, std::ref(the_container)));
		}else if(k < total_deleters){
			deleters.push_back(std::thread(deleter, k++, std::ref(the_container)));
		}
	}
	
	for(auto i = searchers.begin(); i != searchers.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
	for(auto i = inserters.begin(); i != inserters.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
	for(auto i = deleters.begin(); i != deleters.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
}

int main(){
	std::srand(std::time(0));
	try{
		std::cout << "Please input how many searcher threads to run: ";
		int searchers = scan_int();
		if(searchers >= 0){
			std::cout << "Please input how many inserter threads to run: ";
			int inserters = scan_int();
			if(inserters >= 0){
				std::cout << "Please input how many deleter threads to run: ";
				int deleters = scan_int();
				if(deleters >= 0){
					test_scenario(searchers, inserters, deleters);
				}else{
					throw std::invalid_argument("Read a value less than zero from std::cin.");
				}
			}else{
				throw std::invalid_argument("Read a value less than zero from std::cin.");
			}
		}else{
			throw std::invalid_argument("Read a value less than zero from std::cin.");
		}
	}catch(const std::invalid_argument& ex){
		std::cout << "Please input a single integer larger than or equal to two, and nothing else.";
	}
	return 0;
}